### ISSUE TITLE
Mask and QK not of the same shape ?

### ISSUE THREAD

**Issue by Baldwin-disso** [idiap/fast-transformers#101]
I am using fast-transformers encoders in a model and I have
batch size = 16
seq len = 512
n heads = 12

in full_attention.py line 71, QK is of shape (batch, n_heads, seq_len, seq_len) while attention_mask.additive matrix is of shape (batch, seq_len) and of course it can't be summed.

Yet, documentation seems to say that mask is of shape (batch, seq_len) ... 

What Am I missing ? 

regards

**Comment by angeloskath** [comment]
Hi,

Sorry for the late reply.

There are in general three types of masks used throughout the library. `attn_mask`, `query_lengths` and `key_lengths`.

The `attn_mask` always has shape `(seq_len, seq_len)` and encodes where each query can attend to but it is the same for all samples in the batch. The `query_lengths` and `key_lengths` have shape `(batch, seq_len)` and encode which queries and which keys are valid respectively.

So, to sum up, in line 71 the additive matrix has shape `(seq_len, seq_len)` and `QK` has shape `(batch, n_heads, seq_len, seq_len)`) which means the addition is doable with broadcasting.

If you are setting the `attn_mask` manually and you are setting it to shape `(batch, seq_len)` then you will be facing issues. It has been in my to-do list to add a `shape_as()` method in the `BaseMask` interface which would be more readable and could also provide better errors.

I will close the issue but feel free to reopen it if you are having problems.

Cheers,
Angelos
